### ISSUE TITLE
enhancement: Log level verbose

### ISSUE THREAD

**Issue by acrisci** [i3/i3#1823]
Add a `VLOG()` function for verbose logging.

VLOG encodes information into the log line that indicates it is verbose.

    07/30/2015 11:13:38 PM - handlers.c:handle_event:1332 -v- xkb event, need to handle it.

Notice the `-v-` in the middle of the log line. This can be easily detected in a log cleaning utility. Log lines containing this can be filtered easily in `i3-dump-log`.

Certain events that are related to simple input handling (handling xkb events) and all rendering code should use VLOG to indicate it is verbose.

This information is not needed in most of the bugs and features I work on during development.

**Comment by Airblader** [comment]
+1

Especially the rendering code has always been a pain when reading log files.

**Comment by stapelberg** [comment]
Why not just filter out all lines generated by the respective source code files when working with the log, like `grep -v 'render\.c'`?

I’m not convinced we need to change the log format to achieve this effect. Also consider that any change in log format will take quite a while to reach users.

**Comment by Airblader** [comment]
> Why not just filter out all lines generated by the respective source code files

Sure, this is posible. But:

1. This also means I won't see error logs from that file, which usually won't be what I want.
2. IMHO, it's the job of a log level to do basic filtering. That's pretty much what log levels are for. I don't think having to post-process log files is a good idea.
3. Since we usually won't need such trace logs, it also means the log files created by users are much smaller.

As for the actual format, I'd prefer changing all log levels to simply include their log level:

````
[DEBUG] 07/30/2015 11:13:38 PM - foo.c: debug it, baby.
[ERROR] 07/30/2015 11:13:38 PM - foo.c: uh-oh!
[TRACE] 07/30/2015 11:13:38 PM - foo.c: spam me
```

**Comment by stapelberg** [comment]
> This also means I won't see error logs from that file, which usually won't be what I want.

Adjust the grep command line to not trigger when the line contains ERROR :).

> IMHO, it's the job of a log level to do basic filtering. That's pretty much what log levels are for. I don't think having to post-process log files is a good idea.
> Since we usually won't need such trace logs, it also means the log files created by users are much smaller.

I prefer the logs to contain all data and don’t think having them small is a good tradeoff. The main motivation is that small logs are easily achieved using bz2, and having all information inside saves us round-trips to request a log from the user with more details turned on.

**Comment by Airblader** [comment]
> I prefer the logs to contain all data

Then make the default log level trace. If we prefix all logging output as I suggested above, grepping stuff away at least becomes reasonably simple, as would reducing the log level to begin with.

The current output definitely has been dreadful almost every time I had to read a user's log so far, so I really hope we can find a way to improve this.

**Comment by stapelberg** [comment]
> Then make the default log level trace.

Not sure what you mean. Our instructions already make users end up with fully detailed logs?

> If we prefix all logging output as I suggested above, grepping stuff away at least becomes reasonably simple, as would reducing the log level to begin with.

But it also requires categorizing each log message into categories, and it’s unclear to me how these categories differ from the 3 we already have (LOG, ELOG, DLOG). Can you elaborate?

**Comment by acrisci** [comment]
Keep in mind that I'm not suggesting that we log any less for the purposes of the logs we accept for bugs. The actual log file will have all the info that it does now. I'm suggesting that we encode a flag in the log message that can indicate that the message is verbose which can then be used to create different "views" of the log that can be used for different purposes by parsing the log and filtering.

I want to standardize this log processing capability in `i3-dump-log` so anyone can use that to make more sense of the logs.

Here is some example functionality I want from `i3-dump-log`.

    # input a file and show me the errors
    i3-dump-log i3.log --errors

    # input a file and do not show verbose messages (from VLOG)
    i3-dump-log i3.log --log-level debug

    # do not show verbose messages from only certain src files and follow
    i3-dump-log --log-level debug --src handlers.c click.c -f

    # input a file and give me a valid config file from it
    i3-dump-log i3.log --config > ~/.i3/buggy-config

**Comment by stapelberg** [comment]
> I'm suggesting that we encode a flag in the log message that can indicate that the message is verbose which can then be used to create different "views" of the log that can be used for different purposes by parsing the log and filtering.

There are three log levels currently:

```c
#define LOG(fmt, ...) verboselog(fmt, ##__VA_ARGS__)
#define ELOG(fmt, ...) errorlog("ERROR: " fmt, ##__VA_ARGS__)
#define DLOG(fmt, ...) debuglog("%s:%s:%d - " fmt, I3__FILE__, __FUNCTION__, __LINE__, ##__VA_ARGS__)
```

When looking at a log line that is stripped of its date/time information, you can categorize it into each of the different log levels:

1. If it matches the regexp `^[^:]+:[^:]+:\d+ - `, it’s a DLOG line.
2. If it matches the regexp `^ERROR`, it’s an ELOG line.
3. Otherwise, it’s a LOG line.

Can you explain why specifically you need to change the log format, given that one can classify log messages as outlined above?

All of the use-cases you listed seem to be possible without changing the log format:

> input a file and show me the errors
> i3-dump-log i3.log --errors

`grep ' - ERROR' i3.log`

> input a file and do not show verbose messages (from VLOG)
> i3-dump-log i3.log --log-level debug

`grep ' - [^:]*:[^:]*:[0-9]* - ' i3.log`

> do not show verbose messages from only certain src files and follow
> i3-dump-log --log-level debug --src handlers.c click.c -f

`i3-dump-log -f | grep -v ' - handlers.c' | grep -v ' - click.c'`

> input a file and give me a valid config file from it
> i3-dump-log i3.log --config > ~/.i3/buggy-config

`sed -n 's/^.*CONFIG(line\s*[0-9]*): //gp' i3.log`

Now, don’t get me wrong, making this functionality available with a better interface than having to know the log format and use grep/sed is certainly a welcome improvement. I’m on the fence on whether i3-dump-log is the right place to implement that, however. Maybe have a small perl script that wraps i3-dump-log if necessary and processes files otherwise? I’m suggesting doing it in perl because we already use lots of perl.

**Comment by teto** [comment]
I have a similar problem with another program that produces much log. Maybe
it's possible to adopt a standardized logging format already supported by
3rd party programs that could filter & colorize the log output. I haven't
found a good combination log format/log parser yet. Some pointers:
See http://serverfault.com/questions/53894/colorize-monitoring-of-logs or
http://lnav.org/

2015-08-01 13:32 GMT+02:00 Michael Stapelberg <notifications@github.com>:

**Comment by acrisci** [comment]
> There are three log levels currently:

Since LOG is already meant to be a verbose log, can we change the name of the function to VLOG to make it more explicit that this is how it should be used, and then expand its use to other things we consider verbose?

Also, it seems like file filtering won't work with verbose logging (or error logging) as it is now because it doesn't include the name of the file and function. What about changing it to this?

```c
#define VLOG(fmt, ...) verboselog("v - %s:%s:%d - " fmt, I3__FILE__, __FUNCTION__, __LINE__, ##__VA_ARGS__)
#define ELOG(fmt, ...) errorlog("e - %s:%s:%d - " fmt, I3__FILE__, __FUNCTION__, __LINE__, ##__VA_ARGS__)
#define DLOG(fmt, ...) debuglog("%s:%s:%d - " fmt, I3__FILE__, __FUNCTION__, __LINE__, ##__VA_ARGS__)
```

> I’m on the fence on whether i3-dump-log is the right place to implement that, however

I can start working on a perl script for log processing functions, put it in the `contrib` folder, and then I might eventually present that as a replacement to i3-dump-log.

**Comment by Airblader** [comment]
> Since LOG is already meant to be a verbose log, can we change the name of the function to VLOG to make it more explicit that this is how it should be used, and then expand its use to other things we consider verbose?

+1

I'd like to suggest that we use the format `[D]`, `[E]` and so on in the beginning of the line. I believe it's just more common to mark the log level like that. I definitely agree that all log levels should have the same format, that is the log level itself, the source file and line number, the timestamp and the message.

**Comment by stapelberg** [comment]
> Since LOG is already meant to be a verbose log, can we change the name of the function to VLOG to make it more explicit that this is how it should be used, […]

The name of the function `verboselog()` is easy to misunderstand: calling `LOG` is the only way we can communicate anything to the user. The “verbose” in the function name comes from the fact that users normally don’t expect any sort of log from their window manager, hence the need to enable the log at all (`i3 -V`, for verbose). All messages that are passed to `LOG` are expected to make sense to a casual user, and should concern such a user. If users come to us and say “I’ve seen this message in my log, and don’t know what it means or why it appears”, we’ve done it wrong.

> and then expand its use to other things we consider verbose?

Looking at the current `LOG` calls, I’d rather make some of them `DLOG`s than introduce more `LOG` calls. Which ones do you have in mind?

> Also, it seems like file filtering won't work with verbose logging (or error logging) as it is now because it doesn't include the name of the file and function.

I’d argue this is by design. Users (who are the primary consumer of the verbose and error logs) should not have to care about the filename/function/line where the message comes from. Also, the volume of these logs should not be so high as to actually be a nuisance in normal operation/while debugging.

**Comment by Airblader** [comment]
This wouldn't be my understanding. Log output is first and foremost meant to be a readable history of what is going on so that the program execution can be traced. 

Normal users shouldn't have to look into the log output at all and anyone who does should be expected to have a basic form of understanding the syntax and semantics, i.e., know what it means to see a file name and a line number. This seems especially true with i3's focus on power users, but even without that fact I'd see it this way. 

Messages intended for the user should be displayed more prominently, for example via i3-nagbar. Which we do in many cases. 

I'd also not agree that a window manager shouldn't be expected to log. Any application should be expected to log and a window manager is complex enough and central to the user session. 

To summarize, imho the log output should aim to serve those who need a thorough understanding of what is going on, not a newbie friendly prose text that anyone can read and fully understand. And this would typically be contributors and people interested in tracking bugs. Hence, we should try to make it easier for those people to work with the log file.

**Comment by stapelberg** [comment]
> Normal users shouldn't have to look into the log output at all and anyone who does should be expected to have a basic form of understanding the syntax and semantics, i.e., know what it means to see a file name and a line number.

Agreed, I expect people to understand what a source code filename, function name and line number is. That doesn’t mean that the average user would care about any of these, so why print them (for `LOG`, that is)?

> Messages intended for the user should be displayed more prominently, for example via i3-nagbar. Which we do in many cases.

I consider these to be extreme cases. The current calls to `start_nagbar` are for invalid config files and keybindings that cannot be parsed. Both cases are essentially the same (input validation, just at different parts of the program).

> I'd also not agree that a window manager shouldn't be expected to log. Any application should be expected to log and a window manager is complex enough and central to the user session.

So you’re saying that window managers should log by default? Consider that the output of window managers ends up in `~/.xsession-errors` unless explicitly redirected — I don’t think the i3 log ending up in `~/.xsession-errors` is what users expect.

> To summarize, imho the log output should aim to serve those who need a thorough understanding of what is going on, not a newbie friendly prose text that anyone can read and fully understand. And this would typically be contributors and people interested in tracking bugs. Hence, we should try to make it easier for those people to work with the log file.

Why can we not have both, i.e. keep the friendly prose in `LOG`, while at the same time enabling the use-cases for people who work a lot with `DLOG`? Specifically, given that I addressed all use-cases that were presented in this discussion so far, what else is it you need?

**Comment by Airblader** [comment]
> That doesn't mean that the average user would care about any of these, so why print them

Because I don't expect the average user to even look at the log output. Either way, I don't really care about what the average user will see (and am happy to satisfy their needs too). What I care about is what the people actually working with log files will see, which should be a consistent format with all necessary information to ease working with it.

> given that I addressed all use-cases that were presented in this discussion so far, what else is it you need?

I'm just not happy with those solutions. IMHO, we shouldn't have to do weird sed / grep postprocessing on log files. Logging should just be done right to begin with. Personally, I'd be happy as long as

* All log output uses the same format
* All log output states its log level
* All log output states its source file

whereas I don't necessarily require this to be the same output the average user will see; it's what **I'd** like to see in order to work efficiently and comfortably.

Giving the log functions (macros) a better name would be the cherry on top, but I also wouldn't really need that. I do believe that having `ELOG` and `DLOG`, `LOG` sort of seems to be the odd one out because it doesn't really describe what it's for (which apparently caused confusion in the past already as you have noticed).

**Comment by stapelberg** [comment]
> IMHO, we shouldn't have to do weird sed / grep postprocessing on log files. Logging should just be done right to begin with.

Assume all log output would be in the format `level:file:line:number - message`. How would that remove the need to use grep to filter?

Unless I misunderstand what you wrote, I think at this point it boils down to personal preference, i.e. “I like this log format better than the other format”, and we’re not debating any functional change.

**Comment by Airblader** [comment]
> How would that remove the need to use grep to filter?

It wouldn't, but it would remove the need to grep differently for different scenarios. I just really don't see why a consistent log format is something we wouldn't want.

**Comment by stapelberg** [comment]
The advantage of not changing the log format is that we don’t have to make our tools support both formats, and also don’t have to wait until the next major release has reached a sufficiently high percentage of our user base.

That said, given that we only support the most recent major release, maybe making the change, releasing a new version, and then merging the tools (which only operate on the new format) is reasonable. The downside is that it will take much more time than if you were to just write the tools for the current format right away.

What do you think? Can we achieve consensus on what the new log format should look like?

A format I’m familiar with is described at https://google-glog.googlecode.com/svn/trunk/doc/glog.html:
```
F0825 185142 test.cc:22] Check failed: write(1, NULL, 2) >= 0 Write NULL failed: Bad address [14]
```

The first letter is the severity, followed by `MMdd HHmmss file:line] message`.

The advantage is that the format is very compact, but maybe not very clear to users.

**Comment by Airblader** [comment]
> The advantage of not changing the log format is that we don’t have to make our tools support both formats

I'm a little lost – which tools are we talking about? Do we have any kind of log parser?

Given that we only support the latest release anyway, I don't think different formats will haunt as a lot. We can always request something to be retested with the latest version (and usually already do).

> The advantage is that the format is very compact, but maybe not very clear to users.

Personally, I don't care too much as long as the information is present in every single log message and in the same format. This is really the only thing I really care about. On a completely personal opinion, I'd prefer the severity to be a bit more separated. Also, I fear that the datetime format in your example is really hard to read. However, if good tooling exists for that format already, it might be worth it since it avoids us having to write our own tooling.

**Comment by stapelberg** [comment]
> I'm a little lost – which tools are we talking about? Do we have any kind of log parser?

I was referring to the tool(s) which @acrisci proposed to write earlier in this thread.

Do note that https://github.com/i3/i3-github-bot/blob/master/logs.go#L35 actually does match log files to verify people are not uploading random junk.

> On a completely personal opinion, I'd prefer the severity to be a bit more separated. Also, I fear that the datetime format in your example is really hard to read

I agree that the datetime is somewhat hard to read. I think we should use ISO8601 date/time formatting, i.e. yyyy-mm-dd hh:mm:ss. I don’t think we need the timezone offset, since we never compare logs across multiple machines.

> However, if good tooling exists for that format already, it might be worth it since it avoids us having to write our own tooling.

I’m not aware of any tooling for any log format. I think our use-cases are so simple that we can easily cover them by a 100-200 line Perl script. At least that’s the impression I get from the use-cases that we discussed so far.

**Comment by Airblader** [comment]
> I was referring to the tool(s) which @acrisci proposed to write earlier in this thread.

Oh, okay. I don't think we have to write those for any old log format. Even if not every report uses the new format right away, it won't matter since we obviously have been able to deal with it until now as well. :)

> I think we should use ISO8601 date/time formatting, i.e. yyyy-mm-dd hh:mm:ss. I don’t think we need the timezone offset,

Agreed.

> I’m not aware of any tooling for any log format. I think our use-cases are so simple that we can easily cover them by a 100-200 line Perl script

Yeah, I just figured tooling might exist. If it doesn't, it doesn't matter anyway. We certainly have no trouble writing a small script.

**Comment by Airblader** [comment]
I don't actually have an interest in this any longer. How do we want to move forward with this? Do we still want this? @stapelberg @acrisci

**Comment by stapelberg** [comment]
I’m not interested in personally spending any time on this, so if there’s nobody else interested, we can close the issue.

**Comment by Airblader** [comment]
Closing it then for now. @acrisci knows what to do if he is still interested in this. :)
